### ISSUE TITLE
ElasticSearch in 2.3.0rc1, CPU max'ed and 500 error

### ISSUE THREAD

**Issue by cheesegrits** [tootsuite/mastodon#6665]
I installed 2.3.0rc1, uncommented the new es container in the docker-compose, and added the three ES_ lines to the .env.production.

Everything built and restarted, but my CPU has been nailed at 100% ever since (mostly the java processes in ES), and searching on a bare word yields a 500 error, "something went wrong on our end".  Everything else (including @ and # searches) works fine. 

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [x] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

**Comment by cheesegrits** [comment]
I updated to rc2, which seems to have enabled logging to the console, which revealed the problem.  

Caused by: java.nio.file.AccessDeniedException: /usr/share/elasticsearch/data/nodes

So I chown'ed and chgroup'ed ./elasticsearch to uid 1000, and g+rwx'ed it, but I'm still getting that error, which seems to be a known issue:

https://github.com/elastic/elasticsearch-docker/issues/21

... where some folk report still having this problem after changing ownership.

Any suggestions?

**Comment by Gargron** [comment]
From reading that issue, it sounds like the chgrp should've solved it entirely. Have you restarted the container? Maybe recreate the container?

**Comment by cheesegrits** [comment]
You'd have thought the chgrp would do it, but I tried everything.  And from googling around, there's plenty of people have the same issue, even after chgrp'ing.

I finally got it to run by using a named volume

```
  es:
    restart: always
    image: docker.elastic.co/elasticsearch/elasticsearch-oss:6.1.3
    environment:
      - "ES_JAVA_OPTS=-Xms512m -Xmx512m"
    networks:
      - internal_network
### Uncomment to enable ES persistance
    volumes:
      - elasticsearch:/usr/share/elasticsearch/data

volumes:
  elasticsearch:
    driver: local
```

... but I'm still getting an error 500 when searching on plain words in the web app, and seeing this in the console logs:

```
es_1         | [2018-03-08T18:51:41,372][INFO ][o.e.n.Node               ] [_gjn02K] starting ...
es_1         | [2018-03-08T18:51:43,736][INFO ][o.e.t.TransportService   ] [_gjn02K] publish_address {172.19.0.4:9300}, bound_addresses {0.0.0.0:9300}
es_1         | [2018-03-08T18:51:43,980][INFO ][o.e.b.BootstrapChecks    ] [_gjn02K] bound or publishing to a non-loopback address, enforcing bootstrap checks
es_1         | ERROR: [1] bootstrap checks failed
es_1         | [1]: max virtual memory areas vm.max_map_count [65530] is too low, increase to at least [262144]
```

... which appears to be another well known issue with elasticsearch.  I'm working on resolving that now.

**Comment by cheesegrits** [comment]
BTW, one really weird thing with the permissions on the mount binding is that it does create folders and files, but still starts barfing up that AccessDeniedException exception:

```
root@mastodon:~/mastodon# ls -ld elasticsearch/
drwxrwsr-x 3 1000 1000 4096 Mar  8 07:18 elasticsearch/
root@mastodon:~/mastodon# ls -l elasticsearch/
total 4
drwxrwxr-x 3 1000 1000 4096 Mar  8 07:18 nodes
root@mastodon:~/mastodon# ls -l elasticsearch/nodes
total 4
drwxrwxr-x 3 1000 1000 4096 Mar  8 11:23 0
root@mastodon:~/mastodon# ls -l elasticsearch/nodes/0
total 4
-rw-rw-r-- 1 1000 1000    0 Mar  8 07:18 node.lock
drwxrwxr-x 2 1000 1000 4096 Mar  8 11:21 _state
root@mastodon:~/mastodon# ls -l elasticsearch/nodes/0/_state
total 4
-rw-rw-r-- 1 1000 root 71 Mar  8 11:21 node-138.st
```

**Comment by cheesegrits** [comment]
OK, I resolved the vm.max_map_count, on my DigitalOcean Ubuntu Docker droplet (so this will be a fairly common issue for Mastodon instances), as per ...

https://www.elastic.co/guide/en/elasticsearch/reference/5.0/vm-max-map-count.html#vm-max-map-count

... and I no longer see any obvious errors in the log output (when 'up' without -d), although the sheer mass of log output makes it a little hard to grok.

But ... I'm still getting a 500 error ("Something went wrong on our end") when searching on bare words.

Any ideas?

**Comment by cheesegrits** [comment]
The problem seems to be that port 9200 is not being exposed.  I can't curl to it, and the web container is reporting that it can't connect.  It doesn't show any ports for the es container in a "docker ps", and netstat shows nothing on the host on 9200.

```
root@mastodon:~/mastodon# docker ps
CONTAINER ID        IMAGE                                                     COMMAND                  CREATED             STATUS              PORTS                              NAMES
76371c954922        docker.elastic.co/elasticsearch/elasticsearch-oss:6.1.3   "/usr/local/bin/dock…"   12 minutes ago      Up 12 minutes                                          mastodon_es_1
2c72048fdc7f        docker/compose:1.13.0                                     "docker-compose up"      12 minutes ago      Up 12 minutes                                          zealous_chandrasekhar
8ff2bf2231d1        tootsuite/mastodon                                        "/sbin/tini -- bundl…"   14 hours ago        Up 12 minutes       3000/tcp, 4000/tcp                 mastodon_sidekiq_1
544aa45f3721        tootsuite/mastodon                                        "/sbin/tini -- bundl…"   14 hours ago        Up 12 minutes       0.0.0.0:3000->3000/tcp, 4000/tcp   mastodon_web_1
e93a7177c68c        tootsuite/mastodon                                        "/sbin/tini -- yarn …"   14 hours ago        Up 12 minutes       3000/tcp, 0.0.0.0:4000->4000/tcp   mastodon_streaming_1
37bb642082df        redis:4.0-alpine                                          "docker-entrypoint.s…"   3 days ago          Up 12 minutes                                          mastodon_redis_1
04e8ea6f5d66        postgres:9.6-alpine                                       "docker-entrypoint.s…"   3 days ago          Up 12 minutes                                          mastodon_db_1

```

But 'docker log' shows ES thinks it has bound to it:

```
[2018-03-08T20:40:44,010][INFO ][o.e.h.n.Netty4HttpServerTransport] [_gjn02K] publish_address {172.19.0.4:9200}, bound_addresses {0.0.0.0:9200}
```

I've tried dicking around with the yml, adding a 'ports' directive, and http.host and transport.host environment, but so far no luck.

I'll quit spamming now, and just update here if I manage to make any progress.

**Comment by Gargron** [comment]
@cheesegrits Have you uncommented the `es` in `depends_on` in docker-compose.yml for the web and sidekiq containers?

**Comment by cheesegrits** [comment]
I hadn't, no (although sidekiq doesn't have that dependency?).  But I tried, and still got:

```
web_1        | [550d11c4-b7fe-4ceb-8465-bfbcd9c400af] Errno::EADDRNOTAVAIL (Failed to open TCP connection to localhost:9200 (Address not available - connect(2) for "localhost" port 9200)):
```

However, that got me thinking.  As the problem is that whatever I try, I can't get 9200 to be exposed on the host, so I tried adding 'es' as a 'link' on web_1, and changed the ES_HOST in .env.production from 'localhost' to 'es' ... AND IT WORKS!!

I know 'links' is deprecated, but ... I'll take it.

Here's my working docker-compose.yml

```
version: '3'
services:

  db:
    restart: always
    image: postgres:9.6-alpine
    networks:
      - internal_network
## Uncomment to enable DB persistance
    volumes:
      - ./postgres:/var/lib/postgresql/data

  redis:
    restart: always
    image: redis:4.0-alpine
    networks:
      - internal_network
## Uncomment to enable REDIS persistance
    volumes:
      - ./redis:/data

  es:
    restart: always
    image: docker.elastic.co/elasticsearch/elasticsearch-oss:6.1.3
    environment:
      - "ES_JAVA_OPTS=-Xms512m -Xmx512m"
    networks:
      - internal_network
    ports:
      - "9300:9300"
      - "9200:9200"
### Uncomment to enable ES persistance
    volumes:
      - elasticsearch:/usr/share/elasticsearch/data

  web:
    build: .
    image: tootsuite/mastodon
    restart: always
    env_file: .env.production
    command: bundle exec rails s -p 3000 -b '0.0.0.0'
    networks:
      - external_network
      - internal_network
    ports:
      - "3000:3000"
    depends_on:
      - db
      - redis
      - es
    links:
      - es
root@mastodon:~/mastodon# cat docker-compose.yml
version: '3'
services:

  db:
    restart: always
    image: postgres:9.6-alpine
    networks:
      - internal_network
## Uncomment to enable DB persistance
    volumes:
      - ./postgres:/var/lib/postgresql/data

  redis:
    restart: always
    image: redis:4.0-alpine
    networks:
      - internal_network
## Uncomment to enable REDIS persistance
    volumes:
      - ./redis:/data

  es:
    restart: always
    image: docker.elastic.co/elasticsearch/elasticsearch-oss:6.1.3
    environment:
      - "ES_JAVA_OPTS=-Xms512m -Xmx512m"
    networks:
      - internal_network
    ports:
      - "9200:9200"
### Uncomment to enable ES persistance
    volumes:
      - elasticsearch:/usr/share/elasticsearch/data

  web:
    build: .
    image: tootsuite/mastodon
    restart: always
    env_file: .env.production
    command: bundle exec rails s -p 3000 -b '0.0.0.0'
    networks:
      - external_network
      - internal_network
    ports:
      - "3000:3000"
    depends_on:
      - db
      - redis
      - es
    links:
      - es
    volumes:
      - ./public/assets:/mastodon/public/assets
      - ./public/packs:/mastodon/public/packs
      - ./public/system:/mastodon/public/system

  streaming:
    build: .
    image: tootsuite/mastodon
    restart: always
    env_file: .env.production
    command: yarn start
    networks:
      - external_network
      - internal_network
    ports:
      - "4000:4000"
    depends_on:
      - db
      - redis

  sidekiq:
    build: .
    image: tootsuite/mastodon
    restart: always
    env_file: .env.production
    command: bundle exec sidekiq -q default -q mailers -q pull -q push
    depends_on:
      - db
      - redis
    networks:
      - external_network
      - internal_network
    volumes:
      - ./public/packs:/mastodon/public/packs
      - ./public/system:/mastodon/public/system

networks:
  external_network:
  internal_network:
    internal: true

volumes:
  elasticsearch:
    driver: local
```

... and the relevant lines from .env.production ...

```
# Optional ElasticSearch configuration
ES_ENABLED=true
ES_HOST=es
ES_PORT=9200
```

So to summarize, for anyone else struggling:

Run 'docker up' without -d, and see if you have AccessDeniedException exceptions.  If you're ./elasticsearch is already chmod 1000.1000 and group writeable, try changing to a named volume, as per the above yml.

If you see stuff about vm.max_map_count in the logs, do this:

https://www.elastic.co/guide/en/elasticsearch/reference/5.0/vm-max-map-count.html#vm-max-map-count

If you don't see port 9200 in a "docker ps" on the elastsearch container, try the 'links' thing in the yml above, and change ES_HOST in .env.production to 'es'.

@Gargron - thanks for hanging in there with me.  I hope this thread might be of some use to others fighting with this.

**Comment by Gargron** [comment]
You should probably remove `"9200:9200"` from ports of the es declaration though, to not accidentally expose access to it from outside. I'm not sure if links is needed over depends_on, seems like the main problem was that your config .env had "localhost" instead of "es" in it as well as having the depends_on item still commented out.

**Comment by cheesegrits** [comment]
OK, I'll remove the ports line.  Wasn't sure if it was needed.

For the ES_HOST, I was going by the sample env, which has localhost:

https://github.com/tootsuite/mastodon/blob/master/.env.production.sample

I tried it with 'es' as the ES_HOST without links (just depends_on) and web_1 bitched about the name not resolving.  However, at that point I was flailing around blindly, so who knows, I may have changed more than one thing.

Once my instance has settled down (my users have had a rough day) I'll do some more testing and get back to you.  If the 'links' thing proves to be unnecessary, I'll edit my earlier post to reflect that.

But it looks like env.production.sample might need changing.

Thanks again.

**Comment by cheesegrits** [comment]
OK, I tested again, and yup, it works fine without the 'links', just the 'depends_on'.  I must have changed something else when I thought that wasn't working.

So the only real issue with the out-of-box config was the 'localhost' in env.production.sample for ES_HOST.

I tried the bind mounting to ./elasticsearch again, still couldn't get that to work with any combination of permissions - it still happily creates ./elasticsearch/data/nodes and starts to populate it, then starts throwing the access denied exception, so I'm sticking with the named volume.

Updated to RC3, everything else looks good.

**Comment by cheesegrits** [comment]
I see you updated the .env.production.sample, which I think wraps this one up.  Whatever is causing my permissions issues doesn't seem to be systemic, it's just me.  So I'll close this one out.  Thanks for all the work on 2.3.0 and ES.  Much appreciated.
